### ISSUE TITLE
extension_modules config works different on master and slave

### ISSUE THREAD

**Issue by puneetk** [saltstack/salt#20865]
Or i could be wrong, but this is current issue i am having. 

This is the current directory structure we are using:

```
/srv/salt
  /state 
    top.sls
    states for roles of servers 
  /example-states
  /ext_module/
     /_modules
      /_states
      /_renderers
      /pillar
        *py
  /pillar/
```

in /etc/salt/master i have setup extension_modules as follows:

```
extension_modules: srv/salt/ext_module/
```

On the minion I have to setup the extension_modules as follows, which is different from how I understand from the documentation in the sall minion config page. 

```
extension_modules: ../ext_module/
```

**Comment by rallytime** [comment]
I haven't ever used `extension_modules` so I might be wrong, but but it seems like it shouldn't have to be configured on the minion in addition to the master. Either way, this could use some investigation.

**Comment by puneetk** [comment]
We are using it to bucket the ext_pillar and all _modules, _renderers and _states python files. If we keep all the directories under the file_root (/srv/salt) then it works fine,  saltutil.sync_all works and syncs out the files to the minions. If we dont keep them under file_root, then ext_pillar does find the pillar and sync_all doesnt sync out the modules to the minions.

**Comment by puneetk** [comment]
@rallytime i have moved over all the _state|_modules back under file_roots, and left the pillar directory under ext_modules directory. Let me know if this is a bug or works as designed.

**Comment by rallytime** [comment]
@puneetk Honestly, I am not sure. I sort of feel like you shouldn't have to define extension modules at all on the the minion. I think maybe @whiteinge or @thatch45 might know more about the answer here.

**Comment by whiteinge** [comment]
There is currently a disparity between how the Master loads custom modules and how custom modules are distributed and loaded onto Minions.

The `/srv/salt/{_modules,_states,_returners,_grains}` directories are minion-specific. Modules in those dirs are distributed and loaded on Minions. The Master ignores those directories.

If you define the `extension_modules` setting the sub-directories must be named the same as the modules directories in the main Salt source. For example: `/srv/modules/{pillar,runners,modules,states,grains}`. Files in these directories are not sent to minions.

The best pattern to load modules in both places in the current version of Salt is to symlink one directory to the other for each of the module types you care about.

The following issues discuss how best to load the same modules in both places: #13842, #12511.

**Comment by kovacsbalu** [comment]
Hi!
Is there any solution for loading extension_modules from gitfs? We are using some python stuff to fill salt mine_function for get network interface data.

**Comment by whiteinge** [comment]
This should be addressed by #30647.

**Comment by terminalmage** [comment]
@kovacsbalu Can you be more specific? Are you trying to sync custom modules to your minions or to the master?
